File: patient-ios-app/MediHubPatient/Core/Models/EducationModels.swift
import Foundation

// MARK: - Patient education (GET /me/patient/education, PUT .../{id}/progress, GET/POST .../questions)

/// One resource the care team assigned to the patient, joined with their
/// progress (PatientEducationItemDTO). Identified by the resource, not the
/// progress row: the progress table has no unique (patient, resource) row.
struct EducationItem: Decodable, Identifiable {
    let resourceId: String
    let progressId: String?
    let title: String?
    let description: String?
    let resourceType: String?
    let category: String?
    let comprehensionStatus: String?
    let progressPercentage: Int?
    let startedAt: String?
    let completedAt: String?
    let lastAccessedAt: String?
    let rating: Int?
    let feedback: String?
    let needsClarification: Bool?
    let clarificationRequest: String?
    let confirmedUnderstanding: Bool?
    let contentUrl: String?
    let textContent: String?
    let thumbnailUrl: String?
    let videoUrl: String?
    let estimatedDuration: Int?
    let tags: [String]?
    let primaryLanguage: String?
    let isWarningSignContent: Bool?

    var id: String { resourceId }
    /// Finished means a completion stamp, as on the web (`!!completedAt`).
    var isCompleted: Bool { !(completedAt ?? "").isEmpty }
    var hasText: Bool { !(textContent ?? "").isEmpty }
    var hasVideo: Bool { !(videoUrl ?? "").isEmpty }
    var hasLink: Bool { !(contentUrl ?? "").isEmpty }
    var hasContent: Bool { hasText || hasVideo || hasLink }
    var isWarningSign: Bool { isWarningSignContent == true }
}

/// PUT /me/patient/education/{id}/progress (PatientEducationProgressUpdateDTO):
/// every field optional; progress 0...100, rating 1...5.
struct EducationProgressUpdate: Encodable {
    var progressPercentage: Int?
    var rating: Int?
    var feedback: String?
    var confirmedUnderstanding: Bool?
    var needsClarification: Bool?
    var clarificationRequest: String?
}

/// GET/POST /me/patient/education/questions (PatientEducationQuestionResponseDTO).
struct EducationQuestion: Decodable, Identifiable {
    let id: String
    let resourceId: String?
    let questionText: String?
    let isUrgent: Bool?
    let isAnswered: Bool?
    let answerText: String?
    let answeredAt: String?
    let requiresInPersonDiscussion: Bool?
    let appointmentScheduled: Bool?
    let createdAt: String?

    var hasAnswer: Bool { isAnswered == true && !(answerText ?? "").isEmpty }
}

/// PatientEducationQuestionSubmitDTO: questionText 5...2000 characters,
/// resourceId optional (a general question).
struct EducationQuestionSubmit: Encodable {
    let resourceId: String?
    let questionText: String
    let isUrgent: Bool
}

File: patient-android-app/app/src/main/java/com/bitnesttechs/hms/patient/core/models/EducationModels.kt
package com.bitnesttechs.hms.patient.core.models

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/** GET /me/patient/education (PatientEducationItemDTO): a resource assigned to the patient, with their progress. */
@JsonClass(generateAdapter = true)
data class EducationItemDto(
    @Json(name = "resourceId") val resourceId: String,
    @Json(name = "progressId") val progressId: String? = null,
    @Json(name = "title") val title: String? = null,
    @Json(name = "description") val description: String? = null,
    @Json(name = "resourceType") val resourceType: String? = null,
    @Json(name = "category") val category: String? = null,
    @Json(name = "comprehensionStatus") val comprehensionStatus: String? = null,
    @Json(name = "progressPercentage") val progressPercentage: Int? = null,
    @Json(name = "startedAt") val startedAt: String? = null,
    @Json(name = "completedAt") val completedAt: String? = null,
    @Json(name = "lastAccessedAt") val lastAccessedAt: String? = null,
    @Json(name = "rating") val rating: Int? = null,
    @Json(name = "feedback") val feedback: String? = null,
    @Json(name = "needsClarification") val needsClarification: Boolean? = null,
    @Json(name = "clarificationRequest") val clarificationRequest: String? = null,
    @Json(name = "confirmedUnderstanding") val confirmedUnderstanding: Boolean? = null,
    @Json(name = "contentUrl") val contentUrl: String? = null,
    @Json(name = "textContent") val textContent: String? = null,
    @Json(name = "thumbnailUrl") val thumbnailUrl: String? = null,
    @Json(name = "videoUrl") val videoUrl: String? = null,
    @Json(name = "estimatedDuration") val estimatedDuration: Int? = null,
    @Json(name = "tags") val tags: List<String>? = null,
    @Json(name = "primaryLanguage") val primaryLanguage: String? = null,
    @Json(name = "isWarningSignContent") val isWarningSignContent: Boolean? = null
) {
    val isCompleted: Boolean get() = !completedAt.isNullOrBlank()
    val hasContent: Boolean get() = !textContent.isNullOrBlank() || !videoUrl.isNullOrBlank() || !contentUrl.isNullOrBlank()
}

/** PUT /me/patient/education/{id}/progress (PatientEducationProgressUpdateDTO); every field optional. */
@JsonClass(generateAdapter = true)
data class EducationProgressUpdate(
    @Json(name = "progressPercentage") val progressPercentage: Int? = null,
    @Json(name = "rating") val rating: Int? = null,
    @Json(name = "feedback") val feedback: String? = null,
    @Json(name = "confirmedUnderstanding") val confirmedUnderstanding: Boolean? = null,
    @Json(name = "needsClarification") val needsClarification: Boolean? = null,
    @Json(name = "clarificationRequest") val clarificationRequest: String? = null
)

/** GET/POST /me/patient/education/questions (PatientEducationQuestionResponseDTO). */
@JsonClass(generateAdapter = true)
data class EducationQuestionDto(
    @Json(name = "id") val id: String,
    @Json(name = "resourceId") val resourceId: String? = null,
    @Json(name = "questionText") val questionText: String? = null,
    @Json(name = "isUrgent") val isUrgent: Boolean? = null,
    @Json(name = "isAnswered") val isAnswered: Boolean? = null,
    @Json(name = "answerText") val answerText: String? = null,
    @Json(name = "answeredAt") val answeredAt: String? = null,
    @Json(name = "requiresInPersonDiscussion") val requiresInPersonDiscussion: Boolean? = null,
    @Json(name = "appointmentScheduled") val appointmentScheduled: Boolean? = null,
    @Json(name = "createdAt") val createdAt: String? = null
)

/** PatientEducationQuestionSubmitDTO: questionText 5..2000, resourceId optional (a general question). */
@JsonClass(generateAdapter = true)
data class EducationQuestionSubmit(
    @Json(name = "resourceId") val resourceId: String? = null,
    @Json(name = "questionText") val questionText: String,
    @Json(name = "isUrgent") val isUrgent: Boolean = false
)
